docs/docs/optimizing-performance.md

```diff
@@ -0,0 +1,118 @@
+  return this.state.count !== nextState.count;
+}
+```
+
+Keep in mind that React will invoke this function pretty often, so the implementation has to be fast.
+
+## shouldComponentUpdate In Action
+
+Here's a subtree of components. For each one, `SCU` indicates what `shouldComponentUpdate` returned, and `vDOMEq` indicates whether the virtual DOMs were equivalent. Finally, the circle's color indicates whether the component had to be reconciled or not.
+
+<figure><img src="/react/img/docs/should-component-update.png" /></figure>
+
+Since `shouldComponentUpdate` returned `false` for the subtree rooted at C2, React did not generate a new virtual DOM for C2, and React didn't even have to invoke `shouldComponentUpdate` on C4 and C5.
+
+For C1 and C3, `shouldComponentUpdate` returned `true`, so React had to go down to the leaves and check them. For C6 `shouldComponentUpdate` returned `true`, and since the virtual DOMs weren't equivalent React had to reconcile the DOM.
+
+The last interesting case is C8. For this node, React had to compute the virtual DOM, but since it was equal to the old one, it didn't have to reconcile its DOM.
+
+Note that React only had to do DOM mutations for C6, which was inevitable. For C8, it bailed out by comparing the virtual DOMs, and for C2's subtree and C7, it didn't even have to compute the virtual DOM as we bailed out on `shouldComponentUpdate`.
+
+## shouldComponentUpdate Examples
+
+Let's say that you have a component that just renders a string value. We could implement `shouldComponentUpdate` as follows:
+
+```javascript
+class StringComponent extends React.Component {
+  render() {
+    return <div>{this.props.value}</div>;
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    return this.props.value !== nextProps.value;
+  }
+}
+```
+
+If your components use more complex data structures, this might not be so simple. For example, let's say your component renders a comma-separated list of words:
+
+```javascript
+class ListOfWords extends React.Component {
+  render() {
+    return <div>{this.props.words.join(',')}</div>;
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    // We have to do an expensive deep equality check
+    if (props.words.length !== nextProps.words.length) {
+      return true;
+    }
+    for (let i = 0; i < props.words.length; i++) {
+      if (props.words[i] !== nextProps.words[i]) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+```
+
+Since the `ListOfWords` component relies on data deep within the props, we have to to a deep comparison to implement `shouldComponentUpdate`. This may not improve performance at all. The fundamental problem is that we want to quickly check for equality of large data structures, without doing an expensive iteration to compare them.
+
+## The Power Of Immutable Data
+
+[Immutable.js](https://github.com/facebook/immutable-js) solves this problem. It provides immutable, persistent collections that work via structural sharing:
```

Comment:
Can we frame it as "One of the possible solutions"? There are many non-trivial pitfalls associated with Immutable.js, and also many cases where you can use it incorrectly and make the performance worse.

I agree we should promote immutability, but I'd rather have this section focused on treating regular JS structures (object and array) as immutable. For example, show ES6 array spread operator, and then mention [object spread proposal](http://redux.js.org/docs/recipes/UsingObjectSpreadOperator.html).

Like the content of these lessons:

* https://egghead.io/lessons/javascript-redux-avoiding-array-mutations-with-concat-slice-and-spread?course=getting-started-with-redux
* https://egghead.io/lessons/javascript-redux-avoiding-object-mutations-with-object-assign-and-spread?course=getting-started-with-redux

Then we could mention Immutable.js at the bottom as a solution for making deep updates nice, and mention [seamless-immutable](https://github.com/rtfeldman/seamless-immutable) and [immutability-helper](https://github.com/kolodny/immutability-helper) as alternative solutions if you don't want to use Immutable.js.